fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerScanner.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * OnlineIndexerScanner.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.async.MoreAsyncUtil;
+import com.apple.foundationdb.async.RangeSet;
+import com.apple.foundationdb.record.IndexState;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.provider.foundationdb.synchronizedsession.SynchronizedSessionRunner;
+import com.apple.foundationdb.record.query.plan.synthetic.SyntheticRecordFromStoredRecordPlan;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Message;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A base class for different types of online indexing scanners.
```

Comment:
I thought that OnlineIndexerBuilder would have been a little confusing :) 
How would you name this "scan & build" class?

Reply:
FYI: IIRC `OnlineIndexer` actually used to be `OnlineIndexBuilder`, but, then it became confusing to have `OnlineIndexBuilder.Builder`, so it was renamed.

Perhaps `InternalOnlineIndexer` or `OnlineIndexerBase` or `AbstractOnlineIndexer`, both of those are kind of confusing, because, in theory the current `OnlineIndexer` is simply a shell around this class with a bunch of helper methods for various types of scans. I think I like `AbstractOnlineIndexer` the best, although an explanation in javadocs would probably be valuable to explain the difference between `AbstractOnlineIndexer` and `OnlineIndexer`.

Also, Naming is one of the *hard* problems in CS, but if naming is hard, it's sometimes a sign that the abstraction is somewhat off. In this case though, I think it's that the logical name is already taken by `OnlineIndexer`

Reply:
How about `OnlineIndexer` for the API and `Indexer` or `IndexMaker` prefix for the internal implementation classes? 
So it'll be
`IndexerBase`
`IndexerByRecords`
etc.

Reply:
Or `Indexing` - 
`IndexingBase`
`IndexingByRecord`
`IndexingThrottle`
Seems to be more descriptive.

Reply:
I've renamed to `Indexing*`. Note that I hadn't move these classes to a new package because they seem to be using many package-private entities. I hope it'll be ok for now.

Reply:
Yeah, `Indexing*` sounds fine